designs/s3/20200210-design-vscode-s3.md

```diff
@@ -0,0 +1,146 @@
+# UX: S3 explorer
+
+* **Type**: UX Design
+* **Status**: Proposal
+
+## Overview
+
+This document proposes a user experience for interacting with AWS S3 from the
+AWS Toolkit ("the Toolkit") for Visual Studio Code ("VSCode").
+
+Some of the features listed below are vague; they are implemented using VSCode
+or platform conventions analogous to the AWS web console.
+
+## Experience
+
+Each region of the Toolkit explorer tree includes an S3 node (see mockup
+below). The toolkit pulls a list of names and properties of S3 objects in the
+currently active profile and presents this data as a tree in the toolkit
+explorer. This is typical of existing toolkit features such as "Schemas" and
+"CDK".
+
+- When a region is expanded, make an eager request to get the count of S3
+  buckets for that region.
+- Show the count of buckets next to the S3 node.
+- The S3 node is always present. If there are no buckets in the region, "grey
+  out" the S3 node.
+- When the current active profile changes, the tree is not refreshed, but is
+  greyed-out to indicate that it is stale. It is not minimized.
+- When a bucket or object is created or deleted, only that node is added or
+  removed from the tree, _without disturbing the current layout of the tree_.
+
+
+![](20200210-design-vscode-s3-tree.png)
+
+### Create and edit S3 buckets
+
+- Root node context menu (right-click):
+  - Create bucket
+- Bucket node context menu (right-click):
+  - Delete bucket
+  - New folder...
+  - Upload...
+  - Copy name
+  - Copy ARN
+
+### Create and edit S3 objects
+
+- S3 object node context menu (right-click):
+  - Delete object
+  - Copy name
+  - Copy ARN
+  - Copy URL
+  - Upload...
+  - Edit
```

Comment:
or "Open": whatever VSCode calls it when a file is revealed in an editor tab.

Reply:
Is there an opportunity to allow this edit to save changes back to S3? That seems like it would be a very common usecase.

Open, make changes, save.

versus

Open, make changes, find location of this temp file?, reupload.